db-service/lib/SQLService.js

```diff
@@ -152,6 +149,39 @@ class SQLService extends DatabaseService {
     return (await ps.run(values)).changes
   }
 
+  get onDELETE() {
+    return super.onDELETE = cds.env.features.assert_integrity === 'db' ? this.onSIMPLE : deep_delete
+    async function deep_delete(/** @type {Request} */ req) {
+      let { compositions } = req.target
+      if (compositions) {
+        // Transform CQL`DELETE from Foo[p1] WHERE p2` into CQL`DELETE from Foo[p1 and p2]`
+        let { from, where } = req.query.DELETE
+        if (typeof from === 'string') from = { ref: [from] }
+        if (where) {
+          let last = from.ref.at(-1)
+          if (last.where) [ last, where ] = [ last.id, [ { xpr: last.where }, 'and', { xpr: where } ] ]
+          from = {ref:[ ...from.ref.slice(0,-1), { id: last, where }]}
+        }
+        // Process child compositions depth-first
+        let { depth=0, visited=[] } = req
+        visited.push (req.target.name)
```

Comment:
Do we support sth. like 
```cds
entity  E {
  ...
  c1: Composition of X...
  c2: Composition of X...
}
```
the current implementation does not allow parallel compositions to same target.